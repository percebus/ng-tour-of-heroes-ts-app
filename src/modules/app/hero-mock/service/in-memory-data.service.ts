import { Injectable } from '@angular/core';
import { InMemoryDbService } from 'angular-in-memory-web-api';
import { Hero } from '../../hero/types/hero';
import { heroes } from '../data/heroes';

@Injectable({
  providedIn: 'root',
})
export class InMemoryDataService implements InMemoryDbService {
  createDb() {
    return { heroes };
  }

  // Overrides the genId method to ensure that a hero always has an id.
  // If the heroes array is empty,
  // the method below returns the initial number (11).
  // if the heroes array is not empty, the method below returns the highest
  // hero id + 1.
  genId(heroes: Hero[]): number {
    const id =
      heroes.length > 0 ? Math.max(...heroes.map((oHero) => oHero.id)) : 10;

    return id + 1;
  }
}
