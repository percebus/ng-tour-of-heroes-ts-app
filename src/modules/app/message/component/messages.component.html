<div class="messages-component" *ngIf="messageService.messages.length">
  <h2>Messages</h2>
  <button type="button" class="clear" (click)="clear()">Clear messages</button>
  <div *ngFor="let message of messageService.messages">
    {{ message }}
  </div>
</div>
