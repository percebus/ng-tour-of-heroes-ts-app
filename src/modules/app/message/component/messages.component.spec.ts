import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MessagesComponent } from './messages.component';
import { MessageService } from '../service/message.service';

describe('MessagesComponent', () => {
  let oMessageComponent: MessagesComponent;
  let oMessageService: MessageService;
  let oComponentFixture: ComponentFixture<MessagesComponent>;

  describe('w/o messages', () => {
    beforeEach(() => {
      TestBed.configureTestingModule({
        declarations: [MessagesComponent],
      });

      oMessageService = TestBed.inject(MessageService);

      oComponentFixture = TestBed.createComponent(MessagesComponent);
      oMessageComponent = oComponentFixture.componentInstance;

      oComponentFixture.detectChanges();
    });

    it('is instanceof MessageComponent', () => {
      expect(oMessageComponent).toBeInstanceOf(MessagesComponent);
    });

    describe('messages', () => {
      it('is an empty array', () => {
        expect(oMessageComponent.messageService.messages).toEqual([]);
      });
    });

    describe('HTML', () => {
      describe('<app-messages>', () => {
        it('is empty', () => {
          expect(oComponentFixture.nativeElement.children.length).toBe(0);
        });
      });

      it('matches snapshot', () => {
        expect(oComponentFixture).toMatchSnapshot();
      });
    });
  });

  describe('with 1 message', () => {
    beforeEach(() => {
      TestBed.configureTestingModule({
        declarations: [MessagesComponent],
      });

      oMessageService = TestBed.inject(MessageService);

      oComponentFixture = TestBed.createComponent(MessagesComponent);
      oMessageComponent = oComponentFixture.componentInstance;

      oMessageService.add('Message #1');
      oComponentFixture.detectChanges();
    });

    describe('MessageService', () => {
      describe('.messages[]', () => {
        it('contains "Message #1"', () => {
          expect(oMessageComponent.messageService.messages).toEqual(
            ['Message #1']
          );
        });
      });
    });

    describe('HTML', () => {
      describe('<app-messages>', () => {
        it('is NOT empty', () => {
          expect(oComponentFixture.nativeElement.children.length).toBeGreaterThan(0);
        });
      });

      describe('heading', () => {
        it('renders "Messages"', () => {
          const oHTMLElement = oComponentFixture.nativeElement.querySelector(
            '.messages-component h2'
          );

          expect(oHTMLElement.textContent).toEqual('Messages');
        });
      });

      describe('"Clear messages" button', () => {
        it('gets rendered', () => {
          const oHTMLElement = oComponentFixture.nativeElement.querySelector(
            '.messages-component button'
          );

          expect(oHTMLElement.textContent).toEqual('Clear messages');
        });
      });

      it('matches snapshot', () => {
        expect(oComponentFixture).toMatchSnapshot();
      });
    });

    describe('.clear()', () => {
      beforeEach(() => {
        oMessageComponent.clear();
        oComponentFixture.detectChanges();
      });

      it('removes all messages', () => {
        expect(oMessageComponent.messageService.messages).toEqual([]);
      });

      describe('HTML', () => {
        describe('<app-messages>', () => {
          it('is empty', () => {
            expect(oComponentFixture.nativeElement.children.length).toBe(0);
          });
        });

        it('matches snapshot', () => {
          expect(oComponentFixture).toMatchSnapshot();
        });
      });
    });
  });
});
