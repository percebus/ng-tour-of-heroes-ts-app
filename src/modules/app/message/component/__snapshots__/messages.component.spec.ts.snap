// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesComponent w/o messages HTML matches snapshot 1`] = `
<app-messages
  messageService={[Function MessageService]}
>
</app-messages>
`;

exports[`MessagesComponent with 1 message .clear() HTML matches snapshot 1`] = `
<app-messages
  messageService={[Function MessageService]}
>
</app-messages>
`;

exports[`MessagesComponent with 1 message HTML matches snapshot 1`] = `
<app-messages
  messageService={[Function MessageService]}
>
  <div
    class="messages-component"
  >
    <h2>
      Messages
    </h2>
    <button
      class="clear"
      type="button"
    >
      Clear messages
    </button>
    <div>
       Message #1 
    </div>
  </div>
</app-messages>
`;
